core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,247 @@
+    val atomTask = BlacklistAtomTask(stageId, taskIndex)
+    if (!isBlacklistExecutorCacheValid) {
+      reEvaluateExecutorBlacklistAndUpdateCache(sched, atomTask, clock)
+    } else {
+      getExecutorBlacklistFromCache(atomTask).getOrElse {
+        reEvaluateExecutorBlacklistAndUpdateCache(sched, atomTask, clock)
+      }
+    }
+  }
+
+  private def reEvaluateExecutorBlacklistAndUpdateCache(
+      sched: TaskSchedulerImpl,
+      atomTask: BlacklistAtomTask,
+      clock: Clock): Set[String] = {
+    val executors = executorsOnBlacklistedNode(sched) ++
+      strategy.getExecutorBlacklist(executorIdToFailureStatus, atomTask, clock)
+    updateBlacklistExecutorCache(atomTask, executors)
+    executors
+  }
+
+  // The actual implementation is delegated to strategy
+  def nodeBlacklist(): Set[String] = synchronized {
+    if (isBlacklistNodeCacheValid) {
+      getNodeBlacklistFromCache
+    } else {
+      val nodes = strategy.getNodeBlacklist(executorIdToFailureStatus)
+      updateBlacklistNodeCache(nodes)
+      nodes
+    }
+  }
+}
+
+/**
+ * Hide cache details in this trait to make code clean and avoid operation mistake
+ */
+private[scheduler]  trait BlacklistCache {
+
+  // local cache to minimize the the work when query blacklisted executor and node
+  private val blacklistExecutorCache = mutable.HashMap.empty[BlacklistAtomTask, Set[String]]
+  private val blacklistNodeCache = mutable.Set.empty[String]
+
+  // The flag to mark if cache is valid, it will be set to false when executorIdToFailureStatus be
+  // updated and it will be set to true, when called executorBlacklist and nodeBlacklist.
+  private var _isBlacklistExecutorCacheValid : Boolean = false
+  private var _isBlacklistNodeCacheValid: Boolean = false
+
+  private val cacheLock = new Object()
+
+  protected def isBlacklistExecutorCacheValid : Boolean = _isBlacklistExecutorCacheValid
+  protected def isBlacklistNodeCacheValid: Boolean = _isBlacklistNodeCacheValid
+
+  protected def updateBlacklistExecutorCache(
+      atomTask: BlacklistAtomTask,
+      blacklistExecutor: Set[String]): Unit = cacheLock.synchronized {
+    if (!_isBlacklistExecutorCacheValid) {
+      blacklistExecutorCache.clear()
+    }
+    blacklistExecutorCache.update(atomTask, blacklistExecutor)
+    _isBlacklistExecutorCacheValid = true
+  }
+
+  protected def updateBlacklistNodeCache(
+      blacklistNode: Set[String]): Unit = cacheLock.synchronized {
+    if (!_isBlacklistNodeCacheValid) {
+      blacklistNodeCache.clear()
+    }
+    blacklistNodeCache ++= blacklistNode
+    _isBlacklistNodeCacheValid = true
+  }
+
+  protected def invalidAllCache(): Unit = cacheLock.synchronized {
+    _isBlacklistExecutorCacheValid = false
+    _isBlacklistNodeCacheValid = false
+  }
+
+  protected def getExecutorBlacklistFromCache(
+      atomTask: BlacklistAtomTask): Option[Set[String]] = blacklistExecutorCache.get(atomTask)
```

Comment:
super nit:  just my personal perference, if the args go on their own line, can you put the body on its own line as well and wrap w/ braces please?